Support for space separated string

sebdeckers opened the pull request:
This makes it easier to integrate with JWTs where the permissions are stored in a string of space-separated values.

Should otherwise be backwards compatible.

MichielDeMey commented:
Hmm, this does limit the string to use spaces as delimiter, some would make the case that comma separated values are the way to go.

Anyway, I strongly believe that your application should simply handle your internal permissions data format and simply reduce it to an array of strings and pass that on to the middleware so we don't end up with an opinionated data format. (String wise)

It simply feels like we're restricting too much here, I'd like to keep it generic. Any thoughts?

sebdeckers commented:
I believe OAuth scopes are actually defined as space-separated lists.

> The value of the scope parameter is expressed as a list of space-
>    delimited, case-sensitive strings.
https://tools.ietf.org/html/rfc6749#section-3.3

If that's not enough, would it be okay to make the separator an option? (That defaults to `' '`.)

sebdeckers commented:
Sorry if I'm totally off-base here. My current use case is simply a permissions check on the OAuth `'scope'` property of a JWT.

I guess this package could be used as a generic permissions checker, but it makes reference to JWT so this seemed reasonable.

MichielDeMey commented:
Actually, you are correct. This package should support OAuth scopes out-of-the-box as per RFC.

Adding a separator would only introduce unnecessary function arguments, so let's roll with space separated string and document this behavior.

Can you update the README to reflect this change as well?

sebdeckers commented:
FYI ESLint is failing on node.js 0.10 & 0.12 ... Want to downgrade to a compatible version of `standard`?

MichielDeMey commented:
Yes, let's downgrade. It's not mission critical, but the linter needs to pass. :)

sebdeckers commented:
@MichielDeMey I've added a test case and updated the README. Commits squashed.

Looking into the `standard` issue it seems that this is an upstream problem. The version this package is using is `~7.2.1` so things should still work.

> With standard v8.0.0, we are also dropping support for Node.js versions prior to v4. Node.js 0.10 and 0.12 are in maintenance mode and will be unsupported at the end of 2016. Node.js 4 is the current LTS version. If you are using an older version of Node.js, we recommend upgrading to at least Node.js 4 as soon as possible. If you are unable to upgrade to Node.js 4 or higher, then we recommend continuing to use standard v7.x until you are ready to upgrade Node.js.
https://standardjs.com/changelog.html#800---2016-08-23

Can we also drop 0.10 & 0.12 support here, or should we try to fix the problem in standard@7?

sebdeckers commented:
Standard issue traced to this pending fix: https://github.com/evcohen/jsx-ast-utils/pull/32

gillesdemey commented on index.js:
```diff
@@ -48,9 +48,13 @@ Guard.prototype = {
-        return next(new UnauthorizedError('permissions_invalid', {
-          message: 'Permissions should be an Array. Bad format?'
-        }))
+        if (typeof permissions === 'string') {
```
I think makes more sense to check if the input value is a string before we check if the input is an Array so we can parse it and always work with an Array value

```javascript
if (typeof permissions === 'string') {
  permissions = permissions.split(' ')
}
// continue with Array
```

sebdeckers replied:
Agreed, that would be the more intuitive way to write it. Did it this way to maintain the same code path for existing Array usage. Happy to refactor.

gillesdemey replied:
Yes please :) I've also merged your PR to remedy the failing Travis builds so when CI is green we're good to merge and release 🤘

sebdeckers commented:
@gillesdemey Seems to have worked, yay!

MichielDeMey commented:
LGTM! 👍

MichielDeMey commented:
I've gone ahead and merged this, we'll release this soon!

sebdeckers commented:
Thanks for the help, both of you! Drinks next time I'm in Gent. 🍻

MichielDeMey commented:
I just released this in v1.0.0, available from NPM. Thanks for the work everyone!
